src/storage/src/write_batch.rs

```diff
@@ -15,420 +15,159 @@
-    ParseSchema {
-        backtrace: Backtrace,
-        source: datatypes::error::Error,
-    },
-
-    #[snafu(display("Failed to decode, corrupted data {}", message))]
-    DataCorrupted {
-        message: String,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Failed to decode vector, source {}", source))]
-    DecodeVector {
-        backtrace: Backtrace,
-        source: datatypes::error::Error,
-    },
-
-    #[snafu(display("Failed to convert into protobuf struct, source {}", source))]
-    ToProtobuf {
-        source: proto::write_batch::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Failed to convert from protobuf struct, source {}", source))]
-    FromProtobuf {
-        source: proto::write_batch::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display(
-        "Failed to create default value for column {}, source: {}",
-        name,
-        source
-    ))]
-    CreateDefault {
-        name: String,
-        #[snafu(backtrace)]
-        source: datatypes::error::Error,
-    },
-}
+use store_api::storage::{OpType, WriteRequest};
 
-pub type Result<T> = std::result::Result<T, Error>;
+use crate::error::{
+    BatchMissingColumnSnafu, CreateDefaultSnafu, CreateRecordBatchSnafu, Error, HasNullSnafu,
+    LenNotEqualsSnafu, RequestTooLargeSnafu, Result, TypeMismatchSnafu, UnknownColumnSnafu,
+};
 
 /// Max number of updates of a write batch.
-const MAX_BATCH_SIZE: usize = 1_000_000;
+pub(crate) const MAX_BATCH_SIZE: usize = 1_000_000;
+
+/// Data of [WriteBatch].
+///
+/// We serialize this struct to the WAL instead of the whole `WriteBatch` to avoid
+/// storing unnecessary information.
+#[derive(Debug, PartialEq)]
+pub struct Payload {
+    /// Schema of the payload.
+    ///
+    /// This schema doesn't contain internal columns.
+    pub schema: SchemaRef,
+    pub mutations: Vec<Mutation>,
+}
 
-impl ErrorExt for Error {
-    fn status_code(&self) -> StatusCode {
-        StatusCode::InvalidArguments
+impl Payload {
+    /// Creates a new payload with given `schema`.
+    fn new(schema: SchemaRef) -> Payload {
+        Payload {
+            schema,
+            mutations: Vec::new(),
+        }
     }
 
-    fn backtrace_opt(&self) -> Option<&Backtrace> {
-        ErrorCompat::backtrace(self)
+    /// Returns true if there is no mutation in the payload.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.mutations.is_empty()
     }
+}
 
-    fn as_any(&self) -> &dyn Any {
-        self
-    }
+/// A write operation to the region.
+#[derive(Debug, PartialEq)]
+pub struct Mutation {
+    /// Type of the mutation.
+    pub op_type: OpType,
+    /// Data of the mutation.
+    pub record_batch: RecordBatch,
 }
 
 /// Implementation of [WriteRequest].
+#[derive(Debug)]
 pub struct WriteBatch {
-    schema: SchemaRef,
-    mutations: Vec<Mutation>,
-    num_rows: usize,
+    payload: Payload,
+    row_index: usize,
```

Comment:
Prefer the previous name `num_rows`. It's more straightforward to me.

Reply:
I rename it to `num_rows_to_mutate` in #777, what about this name?